[TIP] Add new Threat intelligence plugin

PhilippeOberti opened the pull request:
## Summary

This is the first PR for our [Protection Experience team](https://github.com/orgs/elastic/teams/protections-experience)!! 😄 

This PR adds a new Threat Intelligence (TI) plugin to Kibana. The TI plugin leaves in its own `x-pack/plugins` folder but is loaded within the `security_solution` plugin.

As visible in the video below, the TI plugin has the following features:
- can be accessible from both the old and new navigation
- currently consists of a single Indicators page, that loads Threat Intelligence feed and shows the IOCs in a table
- in case of no Indicators, the main page shows a message suggesting to add Integrations
- each row of the Indicators table has an action column that opens a Flyout component to present the IOC details in a Table and JSON views


https://user-images.githubusercontent.com/17276605/179260943-ec7ee957-7ce0-4d3a-ad9e-9a80b9d63bf7.mov


_Notes: Currently the goal is to have the TI UI visible in Kibana 8.4. If this changes we'll be hiding the entries in the old and new navigations._

Most (if not all) components are unit-tested, have an entry in Storybook.
As we've been using primarily EUI components, accessibility seems to work properly.
i18n has been implemented (hopefully properly as well).
e2e Cypress tests have been written to test the basic functionality.


### How to test

1. checkout branch
2. follow the steps in the `Quick Start` section in the plugin's [README.md](https://github.com/elastic/kibana/pull/136479/files#diff-92d803c07a1e6419728bdf28e2e91d62bc40575a463c8ebd78946f41f96febf7R9)
3. to add data to the Indicators table, follow the steps in the last question in the plugin's [FAQ.md](https://github.com/elastic/kibana/pull/136479/files#diff-8962c2f42895737b96274dbb6f7c4cca66bbbe0a436a65da4251f991278bec1dR11)
4. to load Storybook, run unit tests and e2e test, look at the How to test section at the bottom of the plugin's [CONTRIBUTING.md](https://github.com/elastic/kibana/pull/136479/files#diff-8103a8ac95b7eaf4e2e2401596f6b6dfdf804d07f1db2b98286b95e685d2c8b7R118)


https://github.com/elastic/security-team/issues/4372


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))

lgmys committed:
[TIP] Cypress initialization

lgmys committed:
[TIP] Indicator Of Compromise page

- fetch indicators
- display indicators in data-grid
- create reusable layout component (with header)

https://github.com/elastic/security-team/issues/4242

maxcold committed:
[TIP] Add Empty Page

- show Empty page when no indicators

https://github.com/elastic/security-team/issues/4138

PhilippeOberti committed:
[TIP] Add Indicator view details

- add flyout components to show IOCs details (table and JSON)
- add actions column to Indicator Table to open flyout
- add common full-date formatter

https://github.com/elastic/security-team/issues/4244
https://github.com/elastic/security-team/issues/4245

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/build_plugin_list_docs'

PhilippeOberti committed:
[TIP] Add md files at the root of the plugin

- README with development details
- CONTRIBUTING with PR creation and review details
- FAQ

jbudz reviewed (approved):
storybook/ci configs LGTM

nastasha-solomon commented on x-pack/plugins/threat_intelligence/public/components/empty_page/empty_page.tsx:
```diff
@@ -0,0 +1,103 @@
+
+export const EmptyPage: VFC<EmptyPageProps> = ({ integrationsPageLink }) => {
+  const docsLink =
+    'https://www.elastic.co/guide/en/security/current/es-threat-intel-integrations.html';
```
@gchaps I confirmed with @PhilippeOberti that this link is pointing to the correct Security docs.

lcawl replied:
Rather than having a URL in the code like this (which will not be up-to-date as our content moves), this plugin should use https://github.com/elastic/kibana/tree/main/packages/kbn-doc-links. If you need help, let me know!

nastasha-solomon replied:
@PhilippeOberti when you set up the doc link (as per @lcawl 's suggestion in the previous comment) you can use the `threatIntelInt` keyword to connect to [these threat intel docs](https://www.elastic.co/guide/en/security/current/es-threat-intel-integrations.html).

PhilippeOberti replied:
I updated the empty_page component to read from the `kbn-doc-links` plugin (though the kibana services). I made some small changes to the actual `kbn-doc-links` package (small cleanup and also added the `threatIntelInt` type that was missing in the `types.ts` file. Hopefully that is ok 😄

maxcold replied:
@lcawl, @nastasha-solomon can you review the changes @PhilippeOberti did in kbn-doc-links, the review from @elastic/kibana-docs is required now to be able to merge this ticket

stevewritescode commented on .github/CODEOWNERS:
```diff
@@ -587,6 +587,13 @@ x-pack/plugins/session_view @elastic/awp-platform
 x-pack/plugins/security_solution/public/kubernetes @elastic/awp-platform
 
+## Security Solution sub teams - Protections Experience
+x-pack/plugins/threat_intelligence @elastic/protections-experience
```
@maxcold do you have what you need to get this GitHub team added to the Kibana repo, for future PRs?

maxcold replied:
yes, how do we have it added to the Kibana repo?

maxcold replied:
asked in #kibana slack channel

semd commented on x-pack/plugins/threat_intelligence/public/app.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export const App: VFC<AppProps> = ({ services }) => {
+  return (
+    <KibanaContextProvider services={services}>
```
That's great, I like that you have your own `KibanaContextProvider`, I would only add that instead of passing the `services` from the component consumer (Security), it would probably be better to use the core `services` from your own (threat_intelicence) plugin `start` function. This way you will manage your own services and dependencies instead of relying on the Security dependencies.

PhilippeOberti replied:
that makes thanks, I fixed (hopefully properly), thank you!

semd replied:
Not exactly, the proper way would be to do something like:
```
export const createAppComponent = (services: CoreStart) => {
  const App = () => (
    <KibanaContextProvider services={services}>
      <IndicatorsPage />
    </KibanaContextProvider>
  );
  return App;
};

```
and then in your _x-pack/plugins/threat_intelligence/public/plugin.tsx_, the `start` function would have to pass the services, which are composed of the core services + the plugins that you required in your own kibana.json. It would look like this:
```
  public start(core: CoreStart, plugins: PluginsStart): ThreatIntelligencePluginStart {
    const App = createAppComponent({ ...core, ...plugins });
    return { getComponent: () => App };
  }
```
The `PluginsStart` will have to be defined, I see currently you already have some _requiredPlugins_ in your kibana.json.

It is important to note that the `KibanaContextProvider` merges the parent context services (Security) inside the children's context services (Theat Intel), so, even doing these changes, you'll still have all the Security services available when you do `useKibana().services` inside your context.
As you might guess, this change has no real impact right now. But doing so we decouple the Threat intel dependencies from Security, and you won't have problems if any other solution wants to import your components. Also, if you need to add a plugin dependency that Security does not have (e.g. `charts`), you will be able to add it on your own _Threat Intel_ kibana.json. With the current approach, you would have to add it in the _Security_ kibana.json.

Hope it helps

PhilippeOberti replied:
hmm I think I see, let me give it a shot. Thanks for the detailed answer! 😄

semd commented on x-pack/plugins/security_solution/public/threat_intelligence/links.ts:
```diff
@@ -0,0 +1,31 @@
+      defaultMessage: 'Threat Intelligence',
+    }),
+  ],
+  globalNavOrder: 9005,
```
this one is only needed if `globalNavEnabled: true`

PhilippeOberti replied:
fixed thanks

PhilippeOberti committed:
[TIP] Threat Intelligence initial commit

- create Threat Intelligence plugin and integrate with Security Solution plugin
- update Kibana CODEOWNERS
- setup jest unit tests, i18n and Storybook

https://github.com/elastic/security-team/issues/4329
https://github.com/elastic/security-team/issues/4241

PhilippeOberti committed:
[TIP] Update empty page component to use doc-links package

- add new threatIntelInt entry to kbn-doc-links package
- retrieve link from the threat intelligence empty page component

PhilippeOberti committed:
[TIP] Remove unneeded globalNavOrder in new threat intelligence links object

PhilippeOberti committed:
[TIP] Use core (kibana) services within TI plugin directly instead of passing from SS plugin

maxcold commented on x-pack/plugins/threat_intelligence/public/components/empty_page/empty_page.tsx:
```diff
@@ -22,8 +22,7 @@ interface EmptyPageProps {
 export const EmptyPage: VFC<EmptyPageProps> = ({ integrationsPageLink }) => {
-  const docsLink =
-    'https://www.elastic.co/guide/en/security/current/es-threat-intel-integrations.html';
+  const { docLinks } = useKibana().services;
```
To our discussion yesterday about components vs containers, `EmptyPage` with this addition became a container which is fine, but now looking at the component I'd ask why we have `integrationsPageLink` as a prop but doc link we get from inside the component. what about passing docsLink as a prop the same way and move the logic up either to IndicatorsPage or maybe even better to a new EmptyPage container

PhilippeOberti replied:
yup I thought about it too, changing this now! 😄

PhilippeOberti replied:
I'm not a fan of having an EmptyPage container and an EmptyPage component though. I feel like for each component if we have a container and a component the plugin codebase is going to grow exponentially and it's going to be complicated to navigate...

For me the distinction between components and containers is more from a perspective of _can this component be reused as is_? If we're not planning on reusing the empty page and pass different links to it, why having 2 components (a container and a component)?

PhilippeOberti replied:
for this situation I would be even tempted to move all the logic inside the EmptyPage component (therefore make it a container) and remove everything from the IndicatorsPage...

Anyway we can maybe talk a bit more about it? Also that doesn't have to be done for this PR correct? We could clean up the plugin later?

maxcold replied:
yes, let's discuss. I'm ok with moving the logic inside the EmptyPage component and making it a container as you are right, I don't think it will be reused.

maxcold commented on .github/CODEOWNERS:
```diff
@@ -589,6 +589,13 @@ x-pack/plugins/session_view @elastic/awp-platform
+## Security Solution sub teams - Protections Experience
+x-pack/plugins/threat_intelligence @elastic/protections-experience
+x-pack/plugins/security_solution/public/threat_intelligence @elastic/protections-experience
+x-pack/plugins/security_solution/public/threat_intelligence @elastic/protections-experience
```
duplicated line

PhilippeOberti replied:
good catch, fixed!

PhilippeOberti committed:
[TIP] Update EmptyPage component

- move useIntegrationsPageLink into EmptyPage component
- add new useDocumentationLink hook to EmptyPage component
- fix duplicate in CODEOWNERS
- fix unit tests

PhilippeOberti committed:
[TIP] Fix pluging start logic

- reuse services and plugin dependencies from SS plugin
- clean up kibana.json

semd reviewed (approved):
Security changes LGTM! 🚀 
Thanks @PhilippeOberti

maxcold commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into threat-intelligence-plugin

nastasha-solomon commented on x-pack/plugins/threat_intelligence/public/components/empty_page/empty_page.tsx:
```diff
@@ -0,0 +1,104 @@
+
+export const EmptyPage: VFC = () => {
+  const integrationsPageLink = useIntegrationsPageLink();
+  const documentationLink = useTIDocumentationLink();
```
@PhilippeOberti think `useTIDcoumentationLink` should be `threatIntelInt` (check [here](https://github.com/elastic/kibana/blob/main/packages/kbn-doc-links/src/get_doc_links.ts#L342) for reference). I'll let @lcawl confirm though.

lcawl replied:
It seems to be declared in x-pack/plugins/threat_intelligence/public/hooks/use_documentation_link.tsx, so LGTM

lcawl reviewed (approved):
Doc links LGTM

PhilippeOberti commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into threat-intelligence-plugin

tonymeehan commented:
![image](https://user-images.githubusercontent.com/19255656/182386822-60fa96ab-c308-4c22-9aa6-bfe4c2308055.png)
